src/stores/disk.rs

```diff
@@ -0,0 +1,712 @@
+    }
+
+    /// Return the disk path used
+    pub fn disk_path(&self) -> PathBuf {
+        self.disk_path.clone()
+    }
+
+    fn last_cleanup(connection: &Db) -> SystemTime {
+        match connection.get(LAST_CLEANUP_KEY) {
+            Ok(Some(l)) => match rmp_serde::from_slice::<SystemTime>(&l) {
+                Ok(l) => l,
+                _ => SystemTime::UNIX_EPOCH
+            },
+            _ => SystemTime::UNIX_EPOCH
+        }
+    }
+
+    fn cleanup_expired_entries(connection: &Db) {
+        let now = SystemTime::now();
+        let last_cleanup = DiskCache::<K, V>::last_cleanup(connection);
+
+        if last_cleanup + Duration::from_secs(10) < now {
+            return;
+        }
+
+        for (key, value) in connection.iter().flatten() {
+            if let Ok(cached) = rmp_serde::from_slice::<CachedDiskValue<V>>(&value) {
+                if let Some(expires) = cached.expires {
+                    if now > expires {
+                        let _ = connection.remove(key);
+                    }
+                }
+            }
+        }
+
+        let _ = connection.insert(LAST_CLEANUP_KEY, rmp_serde::to_vec(&SystemTime::now()).unwrap());
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum DiskCacheError {
+    #[error("Storage error")]
+    StorageError(#[from] sled::Error),
+    #[error("Error deserializing cached value")]
+    CacheDeserializtionError(#[from] rmp_serde::decode::Error),
+    #[error("Error serializing cached value")]
+    CacheSerializtionError(#[from] rmp_serde::encode::Error),
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct CachedDiskValue<V> {
+    pub(crate) value: V,
+    pub(crate) expires: Option<SystemTime>,
+    pub(crate) version: Option<u64>,
+}
+
+impl<V> CachedDiskValue<V> {
+    fn new(value: V, expires: Option<SystemTime>) -> Self {
+        Self {
+            value,
+            expires,
+            version: Some(1),
+        }
+    }
+}
+
+impl<K, V> IOCached<K, V> for DiskCache<K, V>
+where
+    K: Display,
+    V: Serialize + DeserializeOwned,
+{
+    type Error = DiskCacheError;
+
+    fn cache_get(&self, key: &K) -> Result<Option<V>, DiskCacheError> {
+        Self::cleanup_expired_entries(&self.connection);
+
+        let key = self.generate_key(key);
+
+        if let Some(data) = self.connection.get(key)? {
+            let cached = rmp_serde::from_slice::<CachedDiskValue<V>>(&data)?;
+
+            if let Some(expires) = cached.expires {
+                if SystemTime::now() < expires {
+                    Ok(Some(cached.value))
+                } else {
+                    Ok(None)
+                }
+            } else {
+                Ok(Some(cached.value))
+            }
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn cache_set(&self, key: K, value: V) -> Result<Option<V>, DiskCacheError> {
+        // TODO: wrap value when serializing to also add expiration data, then add separate thread to clean up the cache?
```

Comment:
I think we should stay away from implicitly spawning any background threads here. Providing a way for users to call some function would be enough